### PR TITLE
Dropdown widget dependent field

### PR DESCRIPTION
**Pull request by jkva**
Hi Andy,

This handles dependencies on fields, with all the logic separated out into `linkspace.js`.
One case I'm not sure I'm handling correctly is the `tree` column type, can you explain what the special case is around this column type so I can account for it?

This PR also sorts the floating data table header issue.

### CODE REVIEW

**Comment by abeverley** [comment]
> One case I'm not sure I'm handling correctly is the tree column type, can you explain what the special case is around this column type so I can account for it?

There is nothing particularly special in it itself, it's just that it doesn't behave like a normal form field so I had to handle it differently (e.g. no jquery change() event). I think now that you've broken everything out nicely, that the tree handler can sit with the rest of the code with the multiselect widget, and not be treated in any special way (other than how the widget itself needs to be interfaced with).

It's also worth noting that a tree can have multiple values, so I loop through each one checking it in turn (see ```get_selected``` in tree.tt). I noticed that for the multiselect widget, that you joined the values together with a space. I realise this is similar to how I was doing it with the old widget, but I actually think it would be best to loop each value in turn, the same as the tree.

Finally, I'm not sure whether this PR is ready for testing yet, but I noticed that ```updateState``` is not being called when a single-value field is changed.

Thanks!

**Comment by jkva** [comment]
> Finally, I'm not sure whether this PR is ready for testing yet, but I noticed that updateState is not being called when a single-value field is changed.

They have their own handlers to distinguish from the multi-option select.

> not be treated in any special way (other than how the widget itself needs to be interfaced with)

Ok, so I'd need to write the code for that.

> It's also worth noting that a tree can have multiple values, so I loop through each one checking it in turn (see get_selected in tree.tt). I noticed that for the multiselect widget, that you joined the values together with a space. I realise this is similar to how I was doing it with the old widget, but I actually think it would be best to loop each value in turn, the same as the tree.

Makes sense. I'll change it accordingly.

**Comment by abeverley** [comment]
>> Finally, I'm not sure whether this PR is ready for testing yet, but I noticed that updateState is not being called when a single-value field is changed.
> 
> They have their own handlers to distinguish from the multi-option select.

Okay. I couldn't seem to get the single-value field to display/hide any dependent fields, and the above was as far as I got with debugging, but I'll try again once you've made the latest changes.

**Comment by jkva** [comment]
https://github.com/ctrlo/GADS/blob/master/views/tree.tt#L68 
^ is this debugging?

https://github.com/ctrlo/GADS/blob/master/views/tree.tt#L77
^ this doesn't seem to follow array-of-objects syntax, and seems incorrect wrt jstree api docs. Am I misreading this?

**Comment by abeverley** [comment]
> https://github.com/ctrlo/GADS/blob/master/views/tree.tt#L68
> ^ is this debugging?

Ooops, yes, thank you. Removed with da118d3

> https://github.com/ctrlo/GADS/blob/master/views/tree.tt#L77
> ^ this doesn't seem to follow array-of-objects syntax, and seems incorrect wrt jstree api docs. Am I misreading this?

You're right, but that code is actually no longer used, so I've just removed it (read-only tree values are now just displayed as text rather than in the widget as read-only).

Thanks!

**Comment by jkva** [comment]
@abeverley how do tree widgets get their values? Are those pulled from an ajax endpoint by the tree JS? I can't find code in the existing templates that generates the tree statically, at least.

**Comment by abeverley** [comment]
> how do tree widgets get their values? Are those pulled from an ajax endpoint by the tree JS?

Yes, it's an AJAX call: https://github.com/ctrlo/GADS/blob/b89df75/views/tree.tt#L59

Endpoint is here: https://github.com/ctrlo/GADS/blob/b89df75/lib/GADS.pm#L1479

**Comment by jkva** [comment]
Tree fields seem to work properly, but I'm running into an issue when
adding a display condition dependency. Is this a regression problem? It
seems the [save] action is missing from display conditions.

On Sat, Jun 2, 2018, at 8:26 PM, Andy Beverley wrote:
>> how do tree widgets get their values? Are those pulled from an ajax
>> endpoint by the tree JS?> Yes, it's an AJAX call:
> https://github.com/ctrlo/GADS/blob/b89df75/views/tree.tt#L59> Endpoint is here:
> https://github.com/ctrlo/GADS/blob/b89df75/lib/GADS.pm#L1479> — You are receiving this because you authored the thread. Reply to
> this email directly, view it on GitHub[1], or mute the thread[2].

Links:

  1. https://github.com/ctrlo/GADS/pull/36#issuecomment-394106955

**Comment by abeverley** [comment]
> Tree fields seem to work properly, but I'm running into an issue when adding a display condition dependency. Is this a regression problem? It seems the [save] action is missing from display conditions.

Yes, I did notice that - I'm not sure how long it's not been working as I didn't test it before. I just updated the database manually for my testing (set ```display_field``` and ```display_regex``` values in the layout table).

I noticed one other problem - I'll add a comment in-line to this PR.

File: views/edit.tt [hunk 1]
```diff
@@ -416,7 +417,12 @@
         [% readonly = "" %]
     [% END %]
 
-    <div class="form-group">
+    <div class="form-group" data-column-id="[% dispcol.id %]" data-column-type="[% dispcol.type %]"
+        [% dependency && 'data-has-dependency="' _ parent.id _ '"' %]
+        [% (parent && dependency) && 'data-dependency="' _ depends.regex_b64 _ '"' %]
```
Comment (abeverley) [inline views/edit.tt]: -        [% (parent && dependency) && 'data-dependency="' _ depends.regex_b64 _ '"' %]
+        [% (parent && dependency) && 'data-dependency="' _ dependency.regex_b64 _ '"' %]

**Pull request merged**